except_star.md

```diff
@@ -1,31 +1,172 @@
+other types in the same `ExceptionGroup` need to be automatically reraised, otherwise it is too easy
+for user code to inadvertently swallow exceptions that it is not handling.
+
+The purpose of this PEP, then, is to add the `except*` syntax for handling `ExceptionGroups`s in
+the interpreter, which in turn requires that `ExceptionGroup` is added as a builtin type. The
+semantics of handling `ExceptionGroup`s are not backwards compatible with the current exception
+handling semantics, so could not modify the behaviour of the `except` keyword and instead added
+the new `except*` syntax.
+
+
+## Specification
+
+### ExceptionGroup
+
+The new builtin exception type, `ExceptionGroup` is a subclass of `BaseException`,
+so it is assignable to `Exception.__cause__` and `Exception.__context__`, and can
+be raised and handled as any exception with `raise ExceptionGroup(...)` and
+`try: ... except ExceptionGroup: ...`.
+
+Its constructor takes two parameters: a message string and a sequence of the nested
+exceptions, for example:
+`ExceptionGroup('many problems', [ValueError('bad value'), TypeError('bad type')])`.
+
+The ExceptionGroup class exposes these parameters in the the fields `msg` and
+`excs` (TODO: did we want to rename excs?).  A nested exception can also be an
+`ExceptionGroup` so the class represents a tree of exceptions (to avoid
+cyclic graph we make the `excs` field immutable?).
+
+The `ExceptionGroup.split()` method gives us a way to extract a from an `ExceptionGroup`
+a subset of the exceptions that satify a certain condition:
+
+```python
+eg = ExceptionGroup("one",
+                    [TypeError(1),
+                    ExceptionGroup("two",
+                                   [TypeError(2), ValueError(3)]])
+
+match, rest = eg.split(lambda e: isinstance(e, TypeError))
+```
 
-* `ExceptionGroup` is an iterable object.
-  E.g. `list(ExceptionGroup(ValueError('a'), TypeError('b')))` is
-  equal to `[ValueError('a'), TypeError('b')]`
+`match` is an exception group with the same structure as `eg`, but with only the `TypeError`
+exceptions:  `ExceptionGroup('one', [TypeError(1), ExceptionGroup('two', [TypeError(2)])])`.
+`rest` is the complement of match: `ExceptionGroup('one', [ExceptionGroup('two', [ValueError(3)])])`.
 
-* `ExceptionGroup` is not an indexable object; essentially
-  it's similar to Python `set`. The motivation for this is that exceptions
-  can occur in random order, and letting users write `group[0]` to access the
-  "first" error is error prone. Although the actual implementation of
-  `ExceptionGroup` will likely use an ordered list of errors to preserve
-  the actual occurrence order for rendering.
+Both `match` and `rest` are newly created exceptions, and the original `eg` is unchanged by the
+`split`. The metadata (cause, context and traceback) is copied from the exception groups of `eg`
+to those derived from the in `match` and `rest`.
```

Comment:
In my current implementation it's iterating over excs, and creating the match/rest lists (recursively). Then it creates the match and rest EGs, which is just to wrap each exc list in a new EG and copy the metadata (or None if the excs list is empty).

We could add a check so that if, say, match excs list is empty then rest is the original EG (and vice versa). It won't save copying the excs into the lists though, I don't think we can easily do that. 

It's here: https://github.com/iritkatriel/cpython/blob/8b6bb5a8b96f70b1ef111fb808fb165865e0f971/Objects/exceptions.c#L797

Reply:
Okay, since you already have a flag/method to avoid creating `rest` in the first place, I think always copying is the safer behavior, so let's stick to that. Maybe clarify that all EGs in the tree/DAG are recursively copied. (Oh, hm, this may clarify that you always treat it as a tree -- in my "diamond" example you'd end up with two copies of `a` in some cases. If we just stick to the tree terminology that's to be expected.)

Reply:
There is a case where nothing is copied - when the whole EG returns True for the predicate function. (Imagine eg.split(ExceptionGroup), or eg.split(lambda: True)).

Reply:
Oh, then that should be documented and it should be fine to keep existing EG objects that are to be included in full.

Reply:
The diamond example is, I think, a case of a malformed EG. Why should it happen organically?

Reply:
Not sure. You could be raising the same EG twice in different tasks, but that sounds fishy too.

I am fine with treating it as a tree, so anything that copies d would end up with two copies of a.